server/aesdsocket.c

```diff
@@ -24,532 +24,637 @@
+if (read_buffer == NULL)
+{
+perror("Read Malloc failed");
+goto exit;
+}
 
-	read_buffer = (char *)malloc(sizeof(char *)*BUFFER_MAX);
-	memset(read_buffer, '\0', BUFFER_MAX);
-	if (read_buffer == NULL)
-	{
-		perror("Read Malloc failed");
-		goto exit;
-	}
+sig_status = sigprocmask(SIG_BLOCK, &x, NULL);
+    if(sig_status == -1)
+    {
+      perror("sig_status - 1");
+    }
 
-	do
-	{
-		//receive data over socket and store it in buffer
-		recv_byte = recv(th_newfd, buffer, sizeof(buffer), 0);
+do
+{
+//receive data over socket and store it in buffer
+recv_byte = recv(th_newfd, buffer, sizeof(buffer), 0);
 
-		//check for new-line character
-		if (!recv_byte || (strchr(buffer, '\n') != NULL))
-			break_loop = 1;
+//check for new-line character
+if (!recv_byte || (strchr(buffer, '\n') != NULL))
+break_loop = 1;
 
-		//check if allocated buffer size is sufficient
-		if ((malloc_size - rdbuff_size) < recv_byte)
-		{
+//check if allocated buffer size is sufficient
+if ((malloc_size - rdbuff_size) < recv_byte)
+{
 
-			malloc_size += recv_byte;
-			//realloc required size
-			read_buffer = (char *)realloc(read_buffer, sizeof(char) * malloc_size);
-		}
+malloc_size += recv_byte;
+//realloc required size
+char *tmp_realloc = realloc(read_buffer, sizeof(char)*malloc_size);
+if(tmp_realloc == NULL) {
+	perror("read_buffer realloc");
+	free(read_buffer);
+	read_buffer = NULL;
+	free(tmp_realloc);
+	tmp_realloc = NULL;
+}
 
-		//copy received buffer in another read_buffer
-		memcpy(&read_buffer[rdbuff_size], buffer, recv_byte);
+}
 
-		rdbuff_size += recv_byte;
+//copy received buffer in another read_buffer
+memcpy(&read_buffer[rdbuff_size], buffer, recv_byte);
 
-	} while (break_loop != 1);
+rdbuff_size += recv_byte;
 
-	break_loop = 0;
+} while (break_loop != 1);
 
-	//get mutex lock for thread-safe writing in output file
-	pthread_mutex_lock(&data_mutex);
+break_loop = 0;
 
-	//write from read_buffer in output file
-	write_byte = write(filefd, read_buffer, rdbuff_size);
-	if (write_byte != rdbuff_size)
-	{
-		perror("File write");
-		goto exit;
-	}
+sig_status = sigprocmask(SIG_UNBLOCK, &x, NULL);
+    if(sig_status == -1)
+    {
+      perror("sig_status - 2");
+    }
 
-	//save the end of file
-	end_pos = lseek(filefd, 0, SEEK_END);
+//get mutex lock for thread-safe writing in output file
+pthread_mutex_lock(&data_mutex);
 
-	//seek the start of the file
-	lseek(filefd, 0, SEEK_SET);
+//write from read_buffer in output file
+write_byte = write(filefd, read_buffer, rdbuff_size);
+if (write_byte != rdbuff_size)
+{
+perror("File write");
+goto exit;
+}
 
-	//release mutex lock after writing to file
-	pthread_mutex_unlock(&data_mutex);
+#ifndef USE_AESD_CHAR_DEVICE
+//save the end of file
+end_pos = lseek(filefd, 0, SEEK_END);
 
-	//execute read and send untill end of file reached
-	while (bytes != end_pos)
-	{
-		wrbuff_size = 0;
-		required_memory = end_pos - cur_pos; //get the required memory size for one line
+#else
+end_pos += write_byte;
```

Comment:
@saloni1307 I think this design assumes you can allocate heap as big as the sum of all bytes for every packet.  It would be better to just read one packet at a time.